torch/ao/quantization/observer.py

```diff
@@ -1131,66 +1131,74 @@ def _non_linear_param_search(self) -> Tuple[torch.Tensor, torch.Tensor]:
         new_max = self.min_val + bin_width * (end_bin + 1)
         return new_min, new_max
 
-    def _adjust_min_max(
-        self, combined_min: torch.Tensor, combined_max: torch.Tensor, upsample_rate: int
-    ) -> Tuple[torch.Tensor, torch.Tensor, int, int]:
-        # We ensure that:
-        # (combined_max - combined_min)/(downsample_rate*Nbins) = (max - min)/(upsample_rate*Nbins)
-        # This allows us to have a common grid of resolution s, where we can align
-        # the input histogram
-        # start_idx maps min_val to the histogram bin index.
-
-        # Compute the width of histogram bins is a straightforward solution, where
-        # hist_bin_width = (self.max_val - self.min_val) / (self.bins * upsample_rate)
-        # Underflow happens if the numerator is close to the smallest positive subnormal number of FP32
-        # Therefore, we avoid such division operation.
-        downsample_rate = int(
-            torch.ceil(
-                ((combined_max - combined_min) / (self.max_val - self.min_val)) * upsample_rate
-            ).item()
+    @staticmethod
+    def upscale_histogram(
+        histogram: torch.Tensor,
+        orig_min: torch.Tensor,
+        orig_max: torch.Tensor,
+        update_min: torch.Tensor,
+        update_max: torch.Tensor,
+        n_bins: int,
+        upsample_rate: int = 4,
+    ):
+        # this turns the histogram into a more fine-coarsed histogram to reduce
+        # bin quantization errors
+        histogram = histogram.repeat_interleave(upsample_rate) / upsample_rate
+        bin_size = (orig_max - orig_min) / (n_bins * upsample_rate)
+        mid_points_histogram = (
+            torch.linspace(orig_min, orig_max, n_bins * upsample_rate + 1)[:-1]
+            + 0.5 * bin_size
+        )
+        new_bin_size = (update_min - update_max) / n_bins
+        boundaries_new_histogram = torch.linspace(update_min, update_max, n_bins)
+        # this maps the mid-poits of the histogram to the new histogram's space
+        bucket_assignments = (
+            torch.bucketize(mid_points_histogram, boundaries_new_histogram) - 1
         )
-        e = downsample_rate / upsample_rate * (self.max_val - self.min_val) - (combined_max - combined_min)
-        start_idx = int(
-            torch.round((self.min_val - combined_min) / (self.max_val - self.min_val) * self.bins * upsample_rate).item()
+        # this then maps the histogram mid-points in the new space, weighted by the original histogram's values
+        # this is just the old histogram in the new histogram's space
+
+        update_histogram = torch.bincount(
+            bucket_assignments, weights=histogram, minlength=n_bins
         )
```

Comment:
this is great! it's really nice that we can use all these existing torch functions to implement this, many of these I saw the first time and yet it all make sense after I look up the docs :)